File: go/testutils/time.go
package testutils

import (
	"time"
)

type MockConstTime struct {
}

func (t *MockConstTime) Now() time.Time {
	return time.Date(2021, 1, 1, 0, 0, 0, 0, time.UTC)
}

type MockStepTime struct {
	step time.Time
}

func (t *MockStepTime) Now() time.Time {
	if t.step.IsZero() {
		t.step = (&MockConstTime{}).Now()
		return t.step
	}
	t.step = t.step.Add(time.Second)
	return t.step
}

File: ts/src/types/time.ts
export abstract class Time {
  abstract Now(): Date;
  abstract Sleep(duration: Duration): Promise<void>;
  TimeSince(start: Date): Duration {
    const now = this.Now();
    return now.getTime() - start.getTime();
  }
}

export type Duration = number;

export enum TimeUnits {
  Microsecond = 1,
  Second = 1000 * Microsecond,
  Minute = 60 * Second,
  Hour = 60 * Minute,
}
